src/codegen/transformations.lisp

```diff
@@ -7,208 +7,200 @@
+      (make-node-while-let
+       :type (node-type node)
+       :label (node-while-let-label node)
+       :pattern (node-while-let-pattern node)
+       :expr (apply traverse (node-while-let-expr node) args)
+       :body (apply traverse (node-while-let-body node) args)))
+    (action (:traverse node-loop node traverse &rest args)
+      (make-node-loop
+       :type (node-type node)
+       :label (node-loop-label node)
+       :body (apply traverse (node-loop-body node) args)))
+    (action (:traverse node-seq node traverse &rest args)
+      (make-node-seq
+       :type (node-type node)
+       :nodes (mapcar
+               (lambda (node)
+                 (apply traverse node args))
+               (node-seq-nodes node))))
+    (action (:traverse node-return node traverse &rest args)
+      (make-node-return
+       :type (node-type node)
+       :expr (apply traverse (node-return-expr node) args)))
+    (action (:traverse node-field node traverse &rest args)
+      (make-node-field
+       :type (node-type node)
+       :name (node-field-name node)
+       :dict (apply traverse (node-field-dict node) args)))
+    (action (:traverse node-dynamic-extent node traverse &rest args)
+      (make-node-dynamic-extent
+       :type (node-type node)
+       :name (node-dynamic-extent-name node)
+       :node (apply traverse (node-dynamic-extent-node node) args)
+       :body (apply traverse (node-dynamic-extent-body node) args)))
+    (action (:traverse node-bind node traverse &rest args)
+      (make-node-bind
+       :type (node-type node)
+       :name (node-bind-name node)
+       :expr (apply traverse (node-bind-expr node) args)
+       :body (apply traverse (node-bind-body node) args))))
+   t))
+
+(defun optional-call (when-key type-key actions traverse args node)
+  "Look up a function in `actions` corresponding to `when-key` and `type-key`, and if it exists, call it with arguments `node`, `traverse` (if `(equal ':traverse (car key))`) and (spread) `args`."
+  (declare (type when-keyword when-key)
+           (type node-subtype type-key)
+           (type action-list  actions)
+           (type function     traverse)
+           (type list         args)
+           (type node         node)
+           (values node &optional))
+  (or
+   (alexandria:when-let*
+       ((action (find-if (lambda (action)
+                           (declare (type action action)
+                                    (values boolean &optional))
+                           (and (equal when-key (action-when action))
+                                (equal type-key (action-type action))))
+                         actions))
+        (f (action-function action)))
+     (if (equal ':traverse when-key)
+         (apply f node traverse args)
+         (apply f node args)))
+   node))
 
-(defun call-if (node key funs bound-variables)
-  (declare (type node node)
-           (type symbol key)
+(defun traverse (initial-node custom-actions &rest initial-args)
```

Comment:
we absolutely need to write a very helpful and perhaps even comprehensive explainer on TRAVERSE, in the doc string or in a comment. We need to show how to use it, with some dead simple examples, so a new coalton contributor can understand it quickly. The traverse+actions is a powerful analysis workhorse, and deserves a lot more than a couple emaciated docstrings.

Depending on how comprehensive, it can even be a .md in docs with a link to it here.